File: hackerrank/algorithms/graph-theory/journey-to-the-moon/journey-to-the-moon-dfs-rec-iter.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# https://www.hackerrank.com/challenges/journey-to-the-moon
# In Ruby. Using iteratively implemented recursive depth-first search.

$VERBOSE = 1

# Adjacency-list representation of an unweighted undirected graph.
class Graph
  def initialize(order)
    @adj = Array.new(order) { [] }
  end

  def add_edge(vertex1, vertex2)
    raise IndexError, 'edge vertex1 is out of range' unless exists?(vertex1)
    raise IndexError, 'edge vertex2 us out of range' unless exists?(vertex2)

    @adj[vertex1] << vertex2
    @adj[vertex2] << vertex1
  end

  # Counts vertices in each component.
  def component_orders
    vis = [false] * @adj.size

    (0...@adj.size)
      .lazy
      .reject { |start| vis[start] }
      .map { |start| dfs(vis, start) }
  end

  # Stack frame for iteratively implementing recursive depth-first search.
  class Frame
    def initialize(row)
      @row = row
      @index = 0
    end

    def next
      return nil if @index == @row.size

      ret = @row[@index]
      @index += 1
      ret
    end
  end
  private_constant :Frame

  private

  def dfs(vis, start)
    raise 'Bug: start vertex already visited' if vis[start]

    vis[start] = true
    stack = [Frame.new(@adj[start])]
    count = 1

    until stack.empty?
      dest = stack[-1].next
      if dest.nil?
        stack.pop
      elsif !vis[dest]
        vis[dest] = true
        count += 1
        stack.push(Frame.new(@adj[dest]))
      end
    end

    count
  end

  def exists?(vertex)
    vertex.between?(0, @adj.size - 1)
  end
end

def read_record
  gets.split.map(&:to_i)
end

def read_graph
  vertex_count, edge_count = read_record
  graph = Graph.new(vertex_count)
  edge_count.times { graph.add_edge(*read_record) }
  graph
end

def count_pairs(cardinalities)
  singles = pairs = 0

  cardinalities.each do |cardinality|
    pairs += singles * cardinality
    singles += cardinality
  end

  pairs
end

puts count_pairs(read_graph.component_orders)

File: hackerrank/algorithms/graph-theory/journey-to-the-moon/journey-to-the-moon-dfs-rec-iter.cpp
// https://www.hackerrank.com/challenges/journey-to-the-moon
// In C++14. Using iteratively implemented recursive depth-first search.

#include <cassert>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <stack>
#include <vector>

namespace {
    // Terminates the program abnormally if the given condition fails.
    inline void ensure(const bool condition) noexcept
    {
        if (!condition) std::abort();
    }

    // Adjacency-list representation of a graph.
    using Graph = std::vector<std::vector<int>>;

    // Visitaton states. (Since std::vector<bool> is a bitset, and weird.)
    enum class Color : bool { white, black };

    // Stack frame for graph DFS.
    struct Frame {
        // The source vertex being traversed from.
        int src;

        // An index into the adjacency-list row of src's forward neighbors.
        int index;
    };

    // Reads a graph as an adjacency list.
    Graph read_graph()
    {
        auto vertex_count = -1, edge_count = -1;
        std::cin >> vertex_count >> edge_count;
        ensure(vertex_count >= 0);
        ensure(edge_count >= 0);

        auto adj = Graph(vertex_count);

        for (; edge_count != 0; --edge_count) {
            auto u = -1, v = -1;
            std::cin >> u >> v;

            ensure(0 <= u && u < vertex_count);
            ensure(0 <= v && v < vertex_count);

            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        return adj;
    }

    // Performs an action for each component size in the graph.
    template<typename UnaryFunction>
    void for_each_component_size(const Graph& adj, UnaryFunction f)
    {
        auto vis = std::vector<Color>(adj.size(), Color::white);
        auto stack = std::stack<Frame>{};

        const auto dfs = [&adj, &vis, &stack](const int start) {
            if (vis[start] != Color::white) return 0;
            vis[start] = Color::black;

            assert(stack.empty());
            stack.push({start, 0});
            auto size = 1;

            while (!stack.empty()) {
                auto& frame = stack.top();
                const auto& row = adj[frame.src];

                if (frame.index == row.size()) {
                    stack.pop();
                    continue;
                }

                const auto dest = row[frame.index++];

                if (vis[dest] == Color::white) {
                    vis[dest] = Color::black;
                    stack.push({dest, 0});
                    ++size;
                }
            }

            return size;
        };

        for (auto start = 0; start != adj.size(); ++start) {
            const auto size = dfs(start);
            if (size != 0) f(size);
        }
    }

    // Computes the number of cross-component unordered vertex pairs.
    std::int_fast64_t count_pairs(const Graph& adj)
    {
        auto total_size = std::int_fast64_t{0};
        auto pair_count = std::int_fast64_t{0}; // May become large.

        const auto do_component = [&total_size, &pair_count](
                const std::int_fast64_t size) noexcept {
            pair_count += total_size * size;
            total_size += size;
        };

        for_each_component_size(adj, do_component);
        return pair_count;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    std::cout << count_pairs(read_graph()) << '\n';
}
